src/cpp/dds/domain/DomainParticipantImpl.cpp

```diff
@@ -0,0 +1,714 @@
+        {
+            std::lock_guard<std::mutex> lock(mtx_pubs_);
+            for (auto pit : publishers_)
+            {
+                pit.second->get_datawriters(writers);
+            }
+        }
+
+        for (DataWriter* writer : writers)
+        {
+            InstanceHandle_t h(writer->guid());
+            if (h == handle)
+            {
+                return true;
+            }
+        }
+
+        // Look into subscribers
+        std::vector<DataReader*> readers;
+        {
+            std::lock_guard<std::mutex> lock(mtx_pubs_);
+            for (auto pit : subscribers_)
+            {
+                pit.second->get_datareaders(readers);
+            }
+        }
+
+        for (DataReader* reader : readers)
+        {
+            InstanceHandle_t h(reader->guid());
+            if (h == handle)
+            {
+                return true;
+            }
+        }
+    }
+
+    return false;
+}
+
+bool DomainParticipantImpl::get_current_time(
+        fastrtps::Time_t& current_time) const
+{
+    auto now = std::chrono::system_clock::now();
+    auto duration = now.time_since_epoch();
+    auto seconds = std::chrono::duration_cast<std::chrono::seconds>(duration);
+    duration -= seconds;
+    auto nanos = std::chrono::duration_cast<std::chrono::nanoseconds>(duration);
+
+    current_time.seconds = static_cast<int32_t>(seconds.count());
+    current_time.nanosec = static_cast<uint32_t>(nanos.count());
+
+    return true;
+}
+
+const DomainParticipant* DomainParticipantImpl::get_participant() const
+{
+    return participant_;
+}
+
+DomainParticipant* DomainParticipantImpl::get_participant()
+{
+    return participant_;
+}
+
+std::vector<std::string> DomainParticipantImpl::get_participant_names() const
+{
+    return rtps_participant_->getParticipantNames();
+}
+
+Subscriber* DomainParticipantImpl::create_subscriber(
+        const SubscriberQos& qos,
+        const fastrtps::SubscriberAttributes& att,
+        SubscriberListener* listen)
+{
+    if(att_.rtps.builtin.discovery_config.use_STATIC_EndpointDiscoveryProtocol)
+    {
+        if(att.getUserDefinedID() <= 0)
+        {
+            logError(PARTICIPANT,"Static EDP requires user defined Id");
+            return nullptr;
+        }
+    }
+
+    if(!att.unicastLocatorList.isValid())
+    {
+        logError(PARTICIPANT,"Unicast Locator List for Subscriber contains invalid Locator");
+        return nullptr;
+    }
+
+    if(!att.multicastLocatorList.isValid())
+    {
+        logError(PARTICIPANT," Multicast Locator List for Subscriber contains invalid Locator");
+        return nullptr;
+    }
+
+    if(!att.remoteLocatorList.isValid())
+    {
+        logError(PARTICIPANT,"Output Locator List for Subscriber contains invalid Locator");
+        return nullptr;
+    }
+
+    if(!att.qos.checkQos() || !att.topic.checkQos())
+    {
+        return nullptr;
+    }
+
+    SubscriberImpl* subimpl = new SubscriberImpl(this, qos, att, listen);
```

Comment:
Add TODO comment like line 219

Reply:
Not a comment from me, but added anyways.